plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1006,98 +1006,108 @@ do it quite yet, even though it worked (the Plutus Tx part wasn't implemented).
+    toBuiltinMeaning _ver AddInteger =
         makeBuiltinMeaning
             ((+) @Integer)
             (runCostingFunTwoArguments . paramAddInteger)
-    toBuiltinMeaning SubtractInteger =
+    toBuiltinMeaning _ver SubtractInteger =
         makeBuiltinMeaning
             ((-) @Integer)
             (runCostingFunTwoArguments . paramSubtractInteger)
-    toBuiltinMeaning MultiplyInteger =
+    toBuiltinMeaning _ver MultiplyInteger =
         makeBuiltinMeaning
             ((*) @Integer)
             (runCostingFunTwoArguments . paramMultiplyInteger)
-    toBuiltinMeaning DivideInteger =
+    toBuiltinMeaning _ver DivideInteger =
         makeBuiltinMeaning
             (nonZeroArg div)
             (runCostingFunTwoArguments . paramDivideInteger)
-    toBuiltinMeaning QuotientInteger =
+    toBuiltinMeaning _ver QuotientInteger =
         makeBuiltinMeaning
             (nonZeroArg quot)
             (runCostingFunTwoArguments . paramQuotientInteger)
-    toBuiltinMeaning RemainderInteger =
+    toBuiltinMeaning _ver RemainderInteger =
         makeBuiltinMeaning
             (nonZeroArg rem)
             (runCostingFunTwoArguments . paramRemainderInteger)
-    toBuiltinMeaning ModInteger =
+    toBuiltinMeaning _ver ModInteger =
         makeBuiltinMeaning
             (nonZeroArg mod)
             (runCostingFunTwoArguments . paramModInteger)
-    toBuiltinMeaning EqualsInteger =
+    toBuiltinMeaning _ver EqualsInteger =
         makeBuiltinMeaning
             ((==) @Integer)
             (runCostingFunTwoArguments . paramEqualsInteger)
-    toBuiltinMeaning LessThanInteger =
+    toBuiltinMeaning _ver LessThanInteger =
         makeBuiltinMeaning
             ((<) @Integer)
             (runCostingFunTwoArguments . paramLessThanInteger)
-    toBuiltinMeaning LessThanEqualsInteger =
+    toBuiltinMeaning _ver LessThanEqualsInteger =
         makeBuiltinMeaning
             ((<=) @Integer)
             (runCostingFunTwoArguments . paramLessThanEqualsInteger)
     -- Bytestrings
-    toBuiltinMeaning AppendByteString =
+    toBuiltinMeaning _ver AppendByteString =
         makeBuiltinMeaning
             BS.append
             (runCostingFunTwoArguments . paramAppendByteString)
-    toBuiltinMeaning ConsByteString =
-        makeBuiltinMeaning
-            (\n xs -> BS.cons (fromIntegral @Integer n) xs)
-            (runCostingFunTwoArguments . paramConsByteString)
-    toBuiltinMeaning SliceByteString =
+    toBuiltinMeaning ver ConsByteString =
+        -- costing-parameter name and costing signature remain the same
+        let costingFun = (runCostingFunTwoArguments . paramConsByteString)
+        -- See Note [Versioned builtins]
+        in case ver of
+            DefaultFunV1 -> makeBuiltinMeaning
+               (\n xs -> BS.cons (fromIntegral @Integer n) xs)
+               costingFun
+            _ -> makeBuiltinMeaning
+              (\(n :: Word8) xs -> BS.cons n xs)
```

Comment:
It might be worth mentioning in a brief comment that there'll be an error if the first argument is out of range.  You have to know quite a lot about unlifting to realise what's going on here, so a reminder might be useful.

Reply:
@kwxm Good idea, AFAIU the range of the input must be between 0-255. 
This reminds me that we need to update the plutuscorespec for v3 with these changes:

ConsByteString  Can Fail YES
with a note that the integer input must be between 0-255

Reply:
@kwxm Do you have a master draft that I can edit already before we forget?

Reply:
>  This reminds me that we need to update the plutuscorespec for v3 with these changes:

I made an issue for that at PLT-629 yesterday.  The main latex source is in the papers repository.  I guess this will be in the next release, so we probably need to add a Chang section.  I'll need to fix the SECP stuff some time too once it's all finalised.

Reply:
> I don't think we should replicate the description of the behavior for every function using `Word8`, `Int` or whatever.

I think this is the first time that we've used `Word8`, and with `Word8` an overflow error is arguably much more likely to show up than with `Int`.  Also the V1 and V2 versions really do have different behaviour here because of that, so you need to pay attention to what's going on.  I don't want to have a big fight about it though!

Reply:
> Also the V1 and V2 versions really do have different behaviour here because of that, so you need to pay attention to what's going on.

That's a very good point.